htlcswitch/circuit.go

```diff
@@ -27,148 +177,547 @@ type PaymentCircuit struct {
+	if err := circuit.Decode(circuitReader); err != nil {
+		return nil, err
+	}
+
+	return circuit, nil
+}
+
+// restoreHalfCircuit reconstructs an in-memory, half payment circuit from a
+// byte slice. The byte slice is assumed to have been generated by the half
+// circuit's Encode method.
+func restoreHalfCircuit(v []byte) (*HalfPaymentCircuit, error) {
+	var halfCircuit = &HalfPaymentCircuit{}
+
+	halfCircuitReader := bytes.NewReader(v)
+	if err := halfCircuit.Decode(halfCircuitReader); err != nil {
+		return nil, err
+	}
+
+	return halfCircuit, nil
 }
 
 // LookupByHTLC looks up the payment circuit by the outgoing channel and HTLC
 // IDs. Returns nil if there is no such circuit.
 func (cm *CircuitMap) LookupByHTLC(chanID lnwire.ShortChannelID, htlcID uint64) *PaymentCircuit {
-	cm.mtx.RLock()
 
-	key := circuitKey{
+	key := CircuitKey{
 		chanID: chanID,
 		htlcID: htlcID,
 	}
-	circuit := cm.circuits[key]
 
-	cm.mtx.RUnlock()
-	return circuit
+	cm.mtx.RLock()
+	defer cm.mtx.RUnlock()
+
+	return cm.circuits[key]
+}
+
+// LookupHalfCircuit looks up a half circuit from the in-memory index. Returns
+// nil if no such half circuit was found.
+func (cm *CircuitMap) LookupHalfCircuit(halfKey CircuitKey) *HalfPaymentCircuit {
+	cm.mtx.RLock()
+	defer cm.mtx.RUnlock()
+
+	return cm.halfCircuits[halfKey]
 }
 
 // LookupByPaymentHash looks up and returns any payment circuits with a given
 // payment hash.
 func (cm *CircuitMap) LookupByPaymentHash(hash [32]byte) []*PaymentCircuit {
 	cm.mtx.RLock()
+	defer cm.mtx.RUnlock()
 
 	var circuits []*PaymentCircuit
 	if circuitSet, ok := cm.hashIndex[hash]; ok {
 		circuits = make([]*PaymentCircuit, 0, len(circuitSet))
-		for circuit := range circuitSet {
-			circuits = append(circuits, &circuit)
+		for key := range circuitSet {
+			if circuit, ok := cm.circuits[key]; ok {
+				circuits = append(circuits, circuit)
+			}
 		}
 	}
 
-	cm.mtx.RUnlock()
 	return circuits
 }
 
-// Add adds a new active payment circuit to the CircuitMap.
-func (cm *CircuitMap) Add(circuit *PaymentCircuit) error {
+// AddHalfCircuit persists the given half circuit to disk and inserts into the
+// in-memory halfCircuits map. HalfAdds will be removed once a call to Add is
+// made for the completed PaymentCircuit.
+func (cm *CircuitMap) AddHalfCircuit(halfCircuit *HalfPaymentCircuit) error {
+	halfKey := halfCircuit.HalfKey()
+
+	// Optimistically serialize the half circuits and half keys.
+	var b bytes.Buffer
+	if err := halfCircuit.Encode(&b); err != nil {
+		return err
+	}
+
 	cm.mtx.Lock()
+	defer cm.mtx.Unlock()
 
-	key := circuitKey{
-		chanID: circuit.OutgoingChanID,
-		htlcID: circuit.OutgoingHTLCID,
+	if _, ok := cm.halfCircuits[halfKey]; ok {
+		return ErrDuplicateHalfCircuitAdd
 	}
-	cm.circuits[key] = circuit
 
-	// Add circuit to the hash index.
-	if _, ok := cm.hashIndex[circuit.PaymentHash]; !ok {
-		cm.hashIndex[circuit.PaymentHash] = make(map[PaymentCircuit]struct{})
+	if err := cm.db.Update(func(tx *bolt.Tx) error {
```

Comment:
Perhaps we should use `Batch` here?

Reply:
And also everywhere else in the PR that we access the `db` in the switch. Possibly we also need to do a sweep over all the channel state machine related access in `channeldb`.